File: problems/hard/bear_and_gene/bear_and_gene.js
function steadyGene(gene) {
      const n = gene.length;
    const required = n / 4;
    const count = { A: 0, C: 0, G: 0, T: 0 };
    for (const c of gene) {
        count[c]++;
    }

    if (Object.values(count).every(c => c === required)) {
        return 0;
    }

    let minLen = n, left = 0;
    for (let right = 0; right < n; right++) {
        count[gene[right]]--;
        while (left < n && Object.values(count).every(c => c <= required)) {
            minLen = Math.min(minLen, right - left + 1);
            count[gene[left]]++;
            left++;
        }
    }
    return minLen;
}

// 🚩 Example usage
if (require.main === module) {
    const readline = require('readline');
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question('Enter gene: ', (gene) => {
        console.log(steadyGene(gene.trim()));
        rl.close();
    });
}

module.exports = steadyGene;

File: problems/hard/bear_and_gene/bear_and_gene.py
def steadyGene(gene):
    n = len(gene)
    required = n // 4
    count = {c: 0 for c in 'ACGT'}
    for c in gene:
        count[c] += 1

    if all(count[c] == required for c in 'ACGT'):
        return 0

    min_len = n
    left = 0
    for right in range(n):
        count[gene[right]] -= 1
        while left < n and all(count[c] <= required for c in 'ACGT'):
            min_len = min(min_len, right - left + 1)
            count[gene[left]] += 1
            left += 1
    return min_len


#  Example usage
if __name__ == '__main__':
    gene = input("Enter the gene string: ").strip()
    print(steadyGene(gene))
